client/protocols/rate_limit.py

```diff
@@ -0,0 +1,81 @@
+from client.protocols.common import models
+from client.exceptions import ProtocolException
+import time
+
+
+BACKOFF_POLICY = (0, 0, 180, 300) # 1-2 no penalty, penalty in seconds after
+
+
+##################
+# Secure Storage #  TODO: refactor persistance to one centralized location
+##################
+
+import shelve # persistance
+import hmac # security
+
+SHELVE_FILE = '.ok_storage'
+SECURITY_KEY = 'uMWm4sviPK3LyPzgWYFn'.encode('utf-8')
+
+def mac(value):
+    mac = hmac.new(SECURITY_KEY)
+    mac.update(repr(value).encode('utf-8'))
+    return mac.hexdigest()
+
+def check(root, key):
+    key = '{}-{}'.format(root, key)
+    with shelve.open(SHELVE_FILE) as db:
+        return key in db
+
+def store(root, key, value):
+    key = '{}-{}'.format(root, key)
+    with shelve.open(SHELVE_FILE) as db:
+        db[key] = {'value': value, 'mac': mac(value)}
```

Comment:
That sounds good. It seems that it is a common theme to get things functional then clean it up a little afterwards :]